src/common/sai.rs

```diff
@@ -0,0 +1,557 @@
+    pub const N: u8 = N;
+
+    /// Create a Sai instance given a set of transmit pins
+    pub fn from_tx(
+        sai: ral::sai::Instance<N>,
+        mut mclk_pin: Mclk,
+        mut tx_pins: Pins<TxSync, TxBclk, TxData>,
+    ) -> Self {
+        reset(&sai);
+
+        sai::prepare(&mut mclk_pin);
+        sai::prepare(&mut tx_pins.sync);
+        sai::prepare(&mut tx_pins.bclk);
+        sai::prepare(&mut tx_pins.data);
+
+        Sai {
+            sai,
+            _mclk_pin: mclk_pin,
+            tx_pins: Some(tx_pins),
+            rx_pins: None,
+            tx_chan_mask: 1 << Chan::to_usize(),
+            rx_chan_mask: 0,
+        }
+    }
+}
+
+impl<const N: u8, Chan, Mclk, RxSync, RxBclk, RxData> Sai<N, Mclk, (), Pins<RxSync, RxBclk, RxData>>
+where
+    Mclk: sai::Pin<consts::Const<N>, Signal = sai::Mclk>,
+    RxSync: sai::Pin<consts::Const<N>, Signal = sai::RxSync>,
+    RxBclk: sai::Pin<consts::Const<N>, Signal = sai::RxBclk>,
+    RxData: sai::Pin<consts::Const<N>>,
+    Chan: consts::Unsigned,
+    <RxData as sai::Pin<consts::Const<N>>>::Signal: sai::RxDataSignal<Index = Chan>,
+{
+    /// Create a Sai instance given a set of receive pins
+    pub fn from_rx(
+        sai: ral::sai::Instance<N>,
+        mut mclk_pin: Mclk,
+        mut rx_pins: Pins<RxSync, RxBclk, RxData>,
+    ) -> Self {
+        reset(&sai);
+
+        sai::prepare(&mut mclk_pin);
+        sai::prepare(&mut rx_pins.sync);
+        sai::prepare(&mut rx_pins.bclk);
+        sai::prepare(&mut rx_pins.data);
+
+        Sai {
+            sai,
+            _mclk_pin: mclk_pin,
+            tx_pins: None,
+            rx_pins: Some(rx_pins),
+            tx_chan_mask: 0,
+            rx_chan_mask: 1 << Chan::to_usize(),
+        }
+    }
+}
+
+impl<const N: u8> Sai<N, (), (), ()> {
+    /// Create a new SAI driver from the RAL SAI instance.
+    ///
+    /// You're responsible for configuring pins, and for making sure
+    /// the pin configuration doesn't change while this driver is in use.
+    /// Setting the channel mask is *also* your responsibility
+    pub fn without_pins(sai: ral::sai::Instance<N>, tx_chan_mask: u32, rx_chan_mask: u32) -> Self {
+        Sai {
+            sai,
+            _mclk_pin: (),
+            tx_pins: None,
+            rx_pins: None,
+            tx_chan_mask,
+            rx_chan_mask,
+        }
+    }
+}
+
+impl<const N: u8, Mclk, TxPins, RxPins> Sai<N, Mclk, TxPins, RxPins> {
+    /// Split the Tx/Rx pair from a SAI, with word, frame, and packing options as type parameters
+    pub fn split<const WORD_SIZE: u8, const FRAME_SIZE: usize, PACKING: Packing<WORD_SIZE>>(
+        self,
+        cfg: &SaiConfig,
+    ) -> (
+        Option<Tx<N, WORD_SIZE, FRAME_SIZE, PACKING>>,
+        Option<Rx<N, WORD_SIZE, FRAME_SIZE, PACKING>>,
+    ) {
+        // Set the mclk pin to be an output
+        unsafe {
+            ral::write_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR1, SAI1_MCLK_DIR: 1);
```

Comment:
Would something like this suffice? Also is this `instance()` call kosher?

```rust
// could this stay in common/sai.rs?
#[cfg(family = "imxrt10xx")]
unsafe fn configure_sai_mclk_dir<const N: u8>(direction: u8) {
    match N {
        1 => ral::modify_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR1, SAI1_MCLK_DIR: direction),
        2 => ral::modify_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR1, SAI2_MCLK_DIR: direction),
        3 => ral::modify_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR1, SAI3_MCLK_DIR: direction),
        _ => panic!("Unsupported SAI instance: {}", N),
    }
}

#[cfg(family = "imxrt11xx")]
unsafe fn configure_sai_mclk_dir<const N: u8>(direction: u8) {
    match N {
        1 => ral::modify_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR0, SAI1_MCLK_DIR: direction),
        2 => ral::modify_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR1, SAI2_MCLK_DIR: direction),
        3 => ral::modify_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR2, SAI3_MCLK_DIR: direction),
        4 => ral::modify_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR2, SAI4_MCLK_DIR: direction),
        _ => panic!("Unsupported SAI instance: {}", N),
    }
}
```

```suggestion
            configure_sai_mclk_dir::<N>(1);
```

Reply:
> Would something like this suffice?

I haven't tested, but I'm not sure if the `family = "imxrt10xx"` branch

```rust
2 => ral::modify_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR1, SAI2_MCLK_DIR: direction),
```

will compile. The 1010 MCUs don't have a `SAI2_MCLK_DIR` field. But don't worry; CI should catch that. If that's the only issue, a conditional compile like

```rust
#[cfg(not(chip="imxrt1010"))]
2 => ral::modify_reg!(ral::iomuxc_gpr, ral::iomuxc_gpr::IOMUXC_GPR::instance(), GPR1, SAI2_MCLK_DIR: direction),
```

might fix it.

Separately, I recommend using the instance validity trick [shown here](https://github.com/imxrt-rs/imxrt-hal/pull/143/commits/06c03ce2df80ae6e8a0caec2b8fa58c7704a2da0) to turn those panics into unreachable branches. A 1010 user might appreciate when their accidental `configure_sai_mclk_dir::<2>(1)` turns into a compile error instead of panicking at runtime.

The input `direction: u8` could be `is_output: bool` to clarify the configuration we're performing. Additionally, a `bool` ensures that we can only pass two values into this one-bit-wide field.

> Also is this instance() call kosher?

That'll have the same chance for race conditions as the previous code. If it's convenient to hide the IOMUXC_GPR  dependency from our users, we can wrap the entire `match` in a [critical_section](https://docs.rs/critical-section/1.1.2/critical_section/).

> could this stay in common/sai.rs?

Let's see how it plays out. If it remains in `common/sai.rs`, I think we'll need to have a `configure_sai_mclk_dir` implementation for when there is no `family` cfg selection. I'm not sure what that implementation would do.